packages/actor-init-query/lib/HttpServiceSparqlEndpoint.ts

```diff
@@ -1,73 +1,344 @@
+        return 'application/n-quads';
+      default:
+        return 'simple';
+    }
+  }
+
+  /**
+   * Extracts the SPARQL protocol operation from an incoming HTTP request.
+   * @param {URL} url The parsed request URL.
+   * @param {IncomingMessage} request The incoming HTTP request.
+   * @returns {ISparqlOperation} The parsed SPARQL protocol operation.
+   */
+  public async parseOperation(url: URL, request: IncomingMessage): Promise<ISparqlOperation> {
+    switch (request.method) {
+      case 'GET':
+      case 'HEAD':
+      case 'OPTIONS':
+        if (url.searchParams.has('query')) {
+          return {
+            type: 'query',
+            queryString: url.searchParams.get('query')!,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        if (url.searchParams.has('update')) {
+          return {
+            type: 'update',
+            queryString: url.searchParams.get('update')!,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        return {
+          type: 'sd',
+          queryString: '',
+          context: this.parseOperationParams(url.searchParams),
+        };
+      case 'POST':
+        // eslint-disable-next-line no-case-declarations
+        const requestBody = await this.readRequestBody(request);
+        if (requestBody.contentType.includes('application/sparql-query')) {
+          return {
+            type: 'query',
+            queryString: requestBody.content,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        if (requestBody.contentType.includes('application/sparql-update')) {
+          return {
+            type: 'update',
+            queryString: requestBody.content,
+            context: this.parseOperationParams(url.searchParams),
+          };
+        }
+        if (requestBody.contentType.includes('application/x-www-form-urlencoded')) {
+          const requestBodyParams = new URLSearchParams(requestBody.content);
+          let requestBodyContext: QueryStringContext | undefined;
+          if (requestBodyParams.has('context')) {
+            try {
+              requestBodyContext = JSON.parse(requestBodyParams.get('context')!);
+            } catch {
+              break;
+            }
+          }
+          if (requestBodyParams.has('query')) {
+            return {
+              type: 'query',
+              queryString: requestBodyParams.get('query')!,
+              context: this.parseOperationParams(url.searchParams, requestBodyContext),
+            };
+          }
+          if (requestBodyParams.has('update')) {
+            return {
+              type: 'update',
+              queryString: requestBodyParams.get('update')!,
+              context: this.parseOperationParams(url.searchParams, requestBodyContext),
+            };
+          }
+        }
+        break;
+      default:
+        throw new HTTPError(405, 'Method Not Allowed');
```

Comment:
Error message is very imprecise. The old error messages were much clearer, e.g. `Invalid POST body with context received...`.